### DISCUSSION TITLE
Thinking About Modularity...

### DISCUSSION THREAD

**Discussion by tajmone** [alan-if/alan-i18n#55] [Library Design]
After having worked at the **Foundation Library** codebase(s) for quite a while now, and drawing from my recent experience with the **ALAN StdLib 2**, I wanted to share some thought regarding what verbs and modules might be worth taking out from the library and turn them into extra optional extension.

# Design Goals of Foundation

The question of how detailed should a "foundational" library be as come a number of times already, and although we never came up with a formalized definition it seems that we all tend to agree that the scope of the **Foundation** is narrower than that of its sibling libraries.

A polite guess of such definitions might be resumed as:

1. It should offer a genre-agnostic adventures template to simplify crafting adventures without having to reinvent the wheel each time.
2. It should avoid opinionated choices and stick to well established IF conventions, yet leaving enough freedom to its author to enforce _their_ opinionated choices without having to struggle.
3. It should not contain verbs which could be deemed as optional in many adventures (score, violence verbs, dressing, etc.).
4. In its built-in messages it should not use a language which suggests a specific genre, mood, epoch/era, or cultural bias (i.e. it should fit the widest context possible).
5. It should be very careful in the choices of its verbs synonyms, and avoid packing under a same syntax verbs which some authors might wish to implement separately (e.g. 'take' and 'grab').
6. The that "less is more" rule should guide most choices.
7. A modular design is ideal for this type of library, but achieving modularity is not that easy and will require some planning.
8. Users should be able to adapt the library to their adventure style without having to alter the library code, so that when they want to use an updated version of the library they don't loose their customization and have to replay them again on the library codebase.

And possibly various other considerations along these lines.

# Externalizing to Extensions

It's important to notice that stripping down the Foundation doesn't mean dumping away its verbs, but simply realizing that many of the current verbs are not strictly essential to every adventure, or are implemented in an opinionated way, and could be moved to separate extension in this project, so that users can still benefit from them by optionally including them in the adventure source.

This would also allow to offer different version of similar extensions.

E.g. the `attack.i` module currently contains various shooting and weapons related verbs. Violence is a sensitive topic, and I see ALAN as a great tool especially for kids, since it's simple to learn and use, so I'd personally like to remove any violence related verbs from the library, into optional extensions.

One of the goals I have in mind, when the Foundation library reaches its first stable release, is to start promoting ALAN as an IF tool for young kids, to be used in schools for educational an fun projects (historical and science simulations, narrative exercises, etc.). But I don't think that including violence related verbs would be a good idea in this case; and the fact that all IF libraries contain the "attack" verb (even if the default answer is "Violence is not the solution") is not a convincing argument — I think that the parser should simply ignore (pretend not to know) verbs which are inappropriate for young kids to use.

At least, all attributes and verbs relating to weapons and firearms should be moved into optional extensions, IMO.

The Foundation should be a _bare-bones_ library, covering _essential_ verbs, classes and grammatical issues, just to simplify authoring.

Furthermore, having multiple choices to pick from via extensions, would become a strong point of the Foundation project, since its very users could share their own extension which they created for their custom adventures — so we could end up with different "weapons extensions": _Wild-West Firearms_, _Sci-Fi Tech Weaponry_, _WW-II Weapons Kit_, _Medieval Arms_, etc. — which sounds more exciting than having to stick to a single opinionated choice, which would also happen to be under-implemented for most use cases (e.g. having firearms without rounds magazines is not very realistic).

# Challenges of Modularization

I've been experimenting with modularization, and haven't yet come up with any easy solution. Unfortunately ALAN doesn't provide a preprocessor so there's no way to branch between alternative code blocks using macros, etc.

The more I look at the Foundation modules, the more I realize that even apparently independent modules in reality depend on certain attributes being defined (e.g. `Is takeable`, etc.). So there are some core modules of the library which define some attributes and/or classes that are needed by other modules, including some that are eligible to be moved into optional extension.

Probably it would be a good idea to try an track which are the core attributes, where they are defined, and which modules depend on them. It's OK if third party and external modules depend on certain core attributes, as long as we are able to preserve these attributes in the course of time, to prevent extensions to break.

So we should identify which are the core modules and attributes, and somehow mark them as being so. Whereas optional modules which could be left out, without affecting other modules, could be marked as being _optional_ modules (regardless of whether we decide to keep them as part of the foundation).

I have the impression that once we start to look at the library from this angle, we'll end up reorganizing its modules in a more efficient way.

# A Few Practical Examples

Take the 'score' meta verb. Not every adventure uses scoring, so the score verb should be in an optional module (i.e. not in `meta_session.i` as it is now). But then, the 'help' verb (`meta_help.i`) lists the various verbs available, including 'score', which means that leaving out the 'score' verb/module would require tweaking the 'help' verb (which is not good).

Probably the best solution is to change how the 'help' verb works, so that each module can "register" its added commands somehow, in order for them to be listed by 'help'. This should actually be quite easily achieved, by exploiting set and string attributes, a hidden library instance, and having each library module add to the set new strings to be shown by 'help'. Such a mechanism would hold true for any third party extension, and authors would have not to worry about tweaking the `meta_help.i` module at all!

Even without macros and preprocessing, it should still be possible to achieve some degree of modularity via some hacks, leveraging hidden library entities to store settings, which would allow authors to control, for example, whether their game will use or not some of the Foundation features.

-------------------------------------------------------------------------------

Anyhow, this is just some food for thoughts which I wanted to share, and I'm sure that everyone who's currently working on a library will have come across similar considerations at some point.

At the moment, keeping these points in mind while working on the libraries code could help focus on the issue, imagine possible solutions, etc.

**Comment by thoni56** [comment]
I think that in the initial incarnations of the library (at least after introduction of v3 concepts like class and additions) the thinking was that a "module", i.e. a file, would `add` the necessary attributes for the verbs in it. Thus an author should be able to copy that particular file and amend it.

I'm not sure how successful or consciously that was executed though. There probably was some cross-over due to not articulating that "principle".

So I feel this line of thought is really good, but have not given the challenges enough thought for many years, and have no immediate ideas or suggestions. Probably, as you say, getting to this mind-set is an important starting point.

**Reply by tajmone** [reply]
> I think that in the initial incarnations of the library (at least after introduction of v3 concepts like class and additions) the thinking was that a "module", i.e. a file, would add the necessary attributes for the verbs in it. Thus an author should be able to copy that particular file and amend it.

I'd say this principle still holds true, but I also believe that present the day ALAN language offers more flexibility in terms of being able to avoid end users having to tweak the library behaviour by altering its source files — a good example is the way @AnssiR66 managed to exploit the `my_game` location (along with a game-start Event) to allow end users to override almost every aspect of the library via the `my_game` instance, leveraging ALAN's order of execution, and exposing library and game settings as attributes on this location.

This trick really opened the door to a novel approach of handling library settings and behaviour customization, reducing to the minimum the instances in which authors are forced to intervene on library source modules, thus allowing them to migrate to newer library versions without having to replay any custom tweaks to its modules — this being the real problem here, i.e. not interfering with updates integration.

> I'm not sure how successful or consciously that was executed though. There probably was some cross-over due to not articulating that "principle".

I noticed that all the **ALAN Lib** modules contained a brief CHANGELOG in the comments at the beginning of the file, reporting all the library releases which affected that particular module, with some brief comments regarding the changes of each version. That was particularly helpful — it surely helped me restoring the latest **ALAN Lib v0.6.2/3** and spotting a mentioned change that somehow went lost in the actual commit to the ALAN source repository.

So we could honour that tradition and come up with a similar system, maybe this time a bit more articulated and also lending itself to info extraction by automated tool, so that they could be easily parsed to track crucial status info via a simple in-comments notation system.

I think that the challenges we're facing today, in this respect, are basically the same: i.e. when an author wants to take advantage of newer library versions, he/she will have to be able to integrate into the new modules any custom tweaks done to the original module (if any), and being able to understand (at a granular and practical level) what has changed and how it might impact the adventure.

It's tempting to propose three-way-diffing as a practical solution for this task, but that would rub against the grain of ALAN's philosophy of preserving simplicity in authorship so that it doesn't leak into the realm of programming expertise. But then, a well-though and clever use of source comments markers in the library could allow us to create a smart diffing tool to help the user in this respect. I have an excellent diffing software that allows the user to manually add custom "alignment markers" to help the tool keep track of diffs between heavily edited documents, especially when whole blocks have been moved up or down in the source.

I have a feeling that will ultimately end up providing some GUI tool with the library anyway, a sort of "Project Wizard" which asks author whether they'll be using certain features in their adventures (e.g. wearables, weapons, dialogs, etc.) and use the gathered info to produce an _ad hoc_ folder with all the required library modules (including optional extensions), the right settings in place, and possibly a boilerplate file for their main adventure source, with some added code and comments to guide the author through the process.

A similar tool could obviate to the lack of preprocessor directives, exploiting in-comments markers in order to manipulate the library modules as required. E.g. the library modules often contain some commented out alternative code, with some brief instructions, which authors can uncomment after deleting the default code, to replace with a common variation (e.g. the `score` verb stating that this game doesn't use score, instead of printing "You've scored 0 out 0 points in _x_ turns", etc.).

A few simple symbols in the comments could guide the tool on where to intervene to achieve the same result. Similarly, the tool would allow users to pick any additional extension and automatically included it in the output folder, along with the required `Import` directives, etc. Nothing rocket-science behind this, but quite helpful on the user-end nonetheless.

I also believe that GUI tools and Wizards to create new projects make new users more comfortable learning ALAN, since they might be scared of using the terminal and command line tools at the beginning. Having a tool that can manage the adventure project, and especially checking for libraries updates and informing the user of which updates would break the adventure and which are safe to download, is quite useful — especially if the tool is able to detect if the original modules where altered somehow (a simple MD5 fingerprinting would do the trick).

Again, I might be abstracting the Issue a lot, but I'm doing so in view of the fact that probably this time we'll have to think of a solution that lends itself to automation (for validation, to start with), but also because I really would like to exploit the Foundation library as a way of promoting ALAN as a tool for school age children. So being able to provide an out-of-the-box set of tools would simplify this, especially when it comes to teachers planning some classes, since simpler tools and project wizards can make a huge difference in terms of how much time-consuming a lesson can become.

But even on simpler level of abstraction, we can clearly see how the introduction of an automated test-suite, and the switch to Rake for handling the toolchain, has benefited maintenance of the library, and how integrating the library code with the documentation and its examples has also improved its quality and the burdens of its development. So I believe that the same is true for handling library updates and its modules: we should take into account the possibility and benefits of tackling the problem from a wider view, not fearing to make it part of the general workflow and practices of the current ecosystem which is slowly but steadily being built around ALAN.

> So I feel this line of thought is really good, but have not given the challenges enough thought for many years, and have no immediate ideas or suggestions. Probably, as you say, getting to this mind-set is an important starting point.

Indeed, I'll attempt to manually sketch a graph of the library dependencies, in terms of how attributes, classes, etc., bind together all the modules, and how these impact modularity in terms of modules which can be simply left out without breaking anything, and those which would require some sort of replacement for the library to keep working. This initial mapping might help figure out which are the "key players", how they are distributed, etc. Then it should be easier to tackle the problem by classifying its constituents and redistributing assets in an optimal way.

But while doing so, this time I'll also be taking into account how in-comments markers could help automate tracking all this info and manipulating modules, which might allow me to see potential solutions where before I could see only dead ends — instead of keep saying, "If only ALAN had a preprocessor", I'll start thinking in terms of "How could an external _ad hoc_ preprocessor help here?"
